File: exercises/python/03_nfa_to_dfa/nfa_to_dfa.py
"""
================================================================================
Exercise 3: NFA to DFA Conversion (Subset Construction / Powerset Construction)
================================================================================

CONCEPT:
    A Non-deterministic Finite Automaton (NFA) can have multiple transitions
    from a single state on the same input symbol, as well as epsilon (empty)
    transitions.  A Deterministic Finite Automaton (DFA) has exactly one
    transition per state per symbol and no epsilon transitions.  Every NFA
    can be converted to an equivalent DFA using the subset construction
    algorithm.

ALGORITHM (Subset Construction):
    1. Compute epsilon-closure(start_state) -- the set of NFA states
       reachable from the start state via epsilon transitions alone.
       This set becomes the DFA start state.
    2. For each unmarked DFA state D (which is a set of NFA states):
       a. Mark D.
       b. For each input symbol 'a':
          - Compute move(D, a) = union of NFA transitions on 'a' from
            every state in D.
          - Compute epsilon-closure(move(D, a)).  This is a new DFA state.
          - If this DFA state is not already in the table, add it.
          - Record the transition D --a--> new_state.
    3. A DFA state is an accepting state if it contains any NFA accepting
       state.

    The epsilon-closure itself is computed by a BFS/DFS from a given set of
    states, following only epsilon edges.

COMPLEXITY:
    Time:  O(2^n * m) in the worst case, where n = number of NFA states and
           m = number of input symbols.  In practice, most DFA states are
           unreachable so the actual number is far smaller.
    Space: O(2^n) for the DFA state table (worst case).

EXAMPLE:
    NFA for (a|b)*abb  ->  DFA with 5 states.
================================================================================
"""

from collections import deque

EPSILON = "eps"

# ---------------------------------------------------------------------------
# NFA representation
# ---------------------------------------------------------------------------
class NFAState:
    """Represents a single NFA state with its transition map."""
    def __init__(self, state_id):
        self.id = state_id
        self.transitions = {}   # symbol -> set of state ids

    def add_transition(self, symbol, target_id):
        self.transitions.setdefault(symbol, set()).add(target_id)


class NFAutomaton:
    """Complete NFA with states, alphabet, start, and accept states."""
    def __init__(self):
        self.states = {}        # id -> NFAState
        self.alphabet = set()   # input symbols (no epsilon)
        self.start = None       # start state id
        self.accept_states = set()

    def add_state(self, state_id):
        if state_id not in self.states:
            self.states[state_id] = NFAState(state_id)

    def add_transition(self, from_id, symbol, to_id):
        self.add_state(from_id)
        self.add_state(to_id)
        self.states[from_id].add_transition(symbol, to_id)
        if symbol != EPSILON:
            self.alphabet.add(symbol)

    def epsilon_closure(self, state_set):
        """Compute the epsilon-closure of a set of NFA state ids."""
        closure = set(state_set)
        worklist = deque(state_set)
        while worklist:
            state_id = worklist.popleft()
            eps_targets = self.states[state_id].transitions.get(EPSILON, set())
            for target in eps_targets:
                if target not in closure:
                    closure.add(target)
                    worklist.append(target)
        return frozenset(closure)

    def move(self, state_set, symbol):
        """Compute move(state_set, symbol): union of transitions on symbol."""
        result = set()
        for state_id in state_set:
            targets = self.states[state_id].transitions.get(symbol, set())
            result |= targets
        return result


# ---------------------------------------------------------------------------
# DFA representation
# ---------------------------------------------------------------------------
class DFAState:
    """A DFA state, which corresponds to a set of NFA states."""
    def __init__(self, dfa_id, nfa_state_set):
        self.id = dfa_id
        self.nfa_states = nfa_state_set   # frozenset of NFA ids
        self.transitions = {}             # symbol -> dfa_id
        self.is_accept = False


class DFAutomaton:
    def __init__(self):
        self.states = {}        # id -> DFAState
        self.start = None
        self.alphabet = set()


# ---------------------------------------------------------------------------
# Subset construction
# ---------------------------------------------------------------------------
def subset_construction(nfa):
    """Convert an NFA to a DFA using the subset construction algorithm."""
    dfa = DFAutomaton()
    dfa.alphabet = set(nfa.alphabet)

    # Map from frozenset of NFA states -> DFA state id
    state_map = {}
    dfa_id_counter = 0

    # 1. Start state = epsilon-closure({nfa.start})
    start_closure = nfa.epsilon_closure({nfa.start})
    state_map[start_closure] = dfa_id_counter
    start_dfa = DFAState(dfa_id_counter, start_closure)
    start_dfa.is_accept = bool(start_closure & nfa.accept_states)
    dfa.states[dfa_id_counter] = start_dfa
    dfa.start = dfa_id_counter
    dfa_id_counter += 1

    # 2. Worklist algorithm
    worklist = deque([start_closure])

    while worklist:
        current_nfa_set = worklist.popleft()
        current_dfa_id = state_map[current_nfa_set]

        for symbol in sorted(dfa.alphabet):
            # move then epsilon-closure
            moved = nfa.move(current_nfa_set, symbol)
            if not moved:
                continue
            new_closure = nfa.epsilon_closure(moved)

            if new_closure not in state_map:
                state_map[new_closure] = dfa_id_counter
                new_dfa = DFAState(dfa_id_counter, new_closure)
                new_dfa.is_accept = bool(new_closure & nfa.accept_states)
                dfa.states[dfa_id_counter] = new_dfa
                dfa_id_counter += 1
                worklist.append(new_closure)

            dfa.states[current_dfa_id].transitions[symbol] = state_map[new_closure]

    return dfa


# ---------------------------------------------------------------------------
# Printing
# ---------------------------------------------------------------------------
def print_nfa(nfa):
    """Print NFA transition table."""
    symbols = [EPSILON] + sorted(nfa.alphabet)
    state_ids = sorted(nfa.states.keys())

    col_w = 18
    state_w = 10
    print(f"{'STATE':<{state_w}}", end="")
    for sym in symbols:
        print(f"| {sym:^{col_w}}", end="")
    print()
    print("-" * (state_w + len(symbols) * (col_w + 2)))

    for sid in state_ids:
        state = nfa.states[sid]
        prefix = ""
        if sid == nfa.start:
            prefix += "->"
        if sid in nfa.accept_states:
            prefix += "*"
        label = f"{prefix}q{sid}"
        print(f"{label:<{state_w}}", end="")
        for sym in symbols:
            targets = state.transitions.get(sym, set())
            cell = ", ".join(f"q{t}" for t in sorted(targets)) if targets else "-"
            print(f"| {cell:^{col_w}}", end="")
        print()

    print("-" * (state_w + len(symbols) * (col_w + 2)))


def print_dfa(dfa):
    """Print DFA transition table with corresponding NFA state sets."""
    symbols = sorted(dfa.alphabet)
    state_ids = sorted(dfa.states.keys())

    col_w = 10
    state_w = 10
    set_w = 24

    print(f"{'STATE':<{state_w}}{'NFA STATES':<{set_w}}", end="")
    for sym in symbols:
        print(f"| {sym:^{col_w}}", end="")
    print()
    print("-" * (state_w + set_w + len(symbols) * (col_w + 2)))

    for sid in state_ids:
        ds = dfa.states[sid]
        prefix = ""
        if sid == dfa.start:
            prefix += "->"
        if ds.is_accept:
            prefix += "*"
        label = f"{prefix}D{sid}"

        nfa_set_str = "{" + ", ".join(f"q{s}" for s in sorted(ds.nfa_states)) + "}"

        print(f"{label:<{state_w}}{nfa_set_str:<{set_w}}", end="")
        for sym in symbols:
            target = ds.transitions.get(sym, None)
            cell = f"D{target}" if target is not None else "-"
            print(f"| {cell:^{col_w}}", end="")
        print()

    print("-" * (state_w + set_w + len(symbols) * (col_w + 2)))


# ---------------------------------------------------------------------------
# Build the example NFA for (a|b)*abb
# ---------------------------------------------------------------------------
def build_example_nfa():
    """
    Build the classic NFA for (a|b)*abb with the following structure:

    State 0 (start) -- eps --> 1, 7
    State 1         -- eps --> 2, 4
    State 2         -- a   --> 3
    State 3         -- eps --> 6
    State 4         -- b   --> 5
    State 5         -- eps --> 6
    State 6         -- eps --> 1, 7
    State 7         -- a   --> 8
    State 8         -- b   --> 9
    State 9         -- b   --> 10 (accept)
    """
    nfa = NFAutomaton()
    nfa.start = 0
    nfa.accept_states = {10}

    transitions = [
        (0, EPSILON, 1), (0, EPSILON, 7),
        (1, EPSILON, 2), (1, EPSILON, 4),
        (2, "a", 3),
        (3, EPSILON, 6),
        (4, "b", 5),
        (5, EPSILON, 6),
        (6, EPSILON, 1), (6, EPSILON, 7),
        (7, "a", 8),
        (8, "b", 9),
        (9, "b", 10),
    ]
    for (src, sym, dst) in transitions:
        nfa.add_transition(src, sym, dst)

    return nfa


def build_nfa_from_user():
    """Interactive NFA input from the user."""
    print("\nEnter NFA details:")
    num_states = int(input("  Number of states: "))
    for i in range(num_states):
        pass  # states 0..n-1

    nfa = NFAutomaton()
    for i in range(num_states):
        nfa.add_state(i)

    nfa.start = int(input("  Start state: "))
    accept_input = input("  Accept states (comma-separated): ")
    nfa.accept_states = set(int(x.strip()) for x in accept_input.split(","))

    num_trans = int(input("  Number of transitions: "))
    print("  Enter each transition as: from symbol to")
    print("  (use 'eps' for epsilon transitions)")
    for _ in range(num_trans):
        parts = input("    > ").split()
        from_id = int(parts[0])
        symbol = parts[1]
        to_id = int(parts[2])
        nfa.add_transition(from_id, symbol, to_id)

    return nfa


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    print("=" * 70)
    print("  NFA -> DFA (Subset Construction)  --  Exercise 3")
    print("=" * 70)

    print("\nChoose input mode:")
    print("  1) Use default NFA for (a|b)*abb")
    print("  2) Enter your own NFA")
    choice = input("\nEnter choice [1]: ").strip()

    if choice == "2":
        nfa = build_nfa_from_user()
    else:
        nfa = build_example_nfa()

    print("\n--- NFA Transition Table ---\n")
    print_nfa(nfa)

    # Convert
    dfa = subset_construction(nfa)

    print("\n--- DFA Transition Table (via Subset Construction) ---\n")
    print_dfa(dfa)

    # Summary
    print(f"\nSummary:")
    print(f"  NFA states: {len(nfa.states)}")
    print(f"  DFA states: {len(dfa.states)}")
    print(f"  Alphabet  : {sorted(dfa.alphabet)}")

    accept_dfa_ids = [sid for sid, ds in dfa.states.items() if ds.is_accept]
    print(f"  DFA start : D{dfa.start}")
    print(f"  DFA accept: {', '.join(f'D{a}' for a in sorted(accept_dfa_ids))}")


if __name__ == "__main__":
    main()

File: exercises/cpp/03_nfa_to_dfa/nfa_to_dfa.cpp
/*
================================================================================
Exercise 3: NFA to DFA Conversion (Subset Construction) -- C++ Implementation
================================================================================

CONCEPT:
    The subset construction (powerset construction) converts an NFA into an
    equivalent DFA.  Each DFA state corresponds to a set of NFA states.  The
    algorithm systematically explores all reachable subsets, computing
    epsilon-closures and move sets to build the DFA transition table.

ALGORITHM:
    1. Compute epsilon-closure({start}).  This becomes DFA state D0.
    2. While there are unmarked DFA states:
       a. Pick an unmarked state D.
       b. For each symbol 'a' in the alphabet:
          - Compute move(D, a) and then epsilon-closure of that.
          - If the result is a new set, add it as a DFA state.
          - Record the DFA transition.
    3. A DFA state is accepting if it contains any NFA accept state.

COMPLEXITY:
    Time:  O(2^n * m) worst case  (n = NFA states, m = alphabet size).
    Space: O(2^n) for DFA states.

EXAMPLE:
    NFA for (a|b)*abb  ->  DFA with 5 states.
================================================================================
*/

#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <map>
#include <queue>
#include <algorithm>
#include <iomanip>
#include <sstream>

/* ---- constants ---------------------------------------------------------- */
static const std::string EPS = "eps";

/* ---- NFA ---------------------------------------------------------------- */
struct NFATransition {
    int         from;
    std::string symbol;
    int         to;
};

class NFA {
public:
    int                         num_states;
    int                         start;
    std::set<int>               accept_states;
    std::set<std::string>       alphabet;   /* excludes epsilon */
    /* transitions indexed by (state, symbol) -> set of target states */
    std::map<std::pair<int,std::string>, std::set<int>> trans;

    NFA() : num_states(0), start(0) {}

    void add_transition(int from, const std::string &symbol, int to) {
        trans[{from, symbol}].insert(to);
        if (symbol != EPS) alphabet.insert(symbol);
        int mx = std::max(from, to) + 1;
        if (mx > num_states) num_states = mx;
    }

    std::set<int> epsilon_closure(const std::set<int> &states) const {
        std::set<int> closure = states;
        std::queue<int> worklist;
        for (int s : states) worklist.push(s);
        while (!worklist.empty()) {
            int curr = worklist.front(); worklist.pop();
            auto it = trans.find({curr, EPS});
            if (it == trans.end()) continue;
            for (int t : it->second) {
                if (closure.insert(t).second)
                    worklist.push(t);
            }
        }
        return closure;
    }

    std::set<int> move(const std::set<int> &states, const std::string &symbol) const {
        std::set<int> result;
        for (int s : states) {
            auto it = trans.find({s, symbol});
            if (it != trans.end())
                result.insert(it->second.begin(), it->second.end());
        }
        return result;
    }
};

/* ---- DFA ---------------------------------------------------------------- */
struct DFAState {
    int                 id;
    std::set<int>       nfa_states;
    bool                is_accept;
    std::map<std::string, int> transitions;
};

class DFA {
public:
    std::vector<DFAState>      states;
    int                        start;
    std::set<std::string>      alphabet;
};

/* ---- subset construction ------------------------------------------------ */
DFA subset_construction(const NFA &nfa) {
    DFA dfa;
    dfa.alphabet = nfa.alphabet;

    std::map<std::set<int>, int> state_map;
    int id_counter = 0;

    /* start state */
    std::set<int> start_set = nfa.epsilon_closure({nfa.start});
    state_map[start_set] = id_counter;
    {
        DFAState ds;
        ds.id = id_counter++;
        ds.nfa_states = start_set;
        ds.is_accept = false;
        for (int s : start_set)
            if (nfa.accept_states.count(s)) { ds.is_accept = true; break; }
        dfa.states.push_back(ds);
    }
    dfa.start = 0;

    std::queue<std::set<int>> worklist;
    worklist.push(start_set);

    while (!worklist.empty()) {
        std::set<int> current = worklist.front(); worklist.pop();
        int curr_id = state_map[current];

        for (auto &sym : dfa.alphabet) {
            std::set<int> moved = nfa.move(current, sym);
            if (moved.empty()) continue;
            std::set<int> closure = nfa.epsilon_closure(moved);
            if (closure.empty()) continue;

            if (state_map.find(closure) == state_map.end()) {
                state_map[closure] = id_counter;
                DFAState ds;
                ds.id = id_counter++;
                ds.nfa_states = closure;
                ds.is_accept = false;
                for (int s : closure)
                    if (nfa.accept_states.count(s)) { ds.is_accept = true; break; }
                dfa.states.push_back(ds);
                worklist.push(closure);
            }
            dfa.states[curr_id].transitions[sym] = state_map[closure];
        }
    }

    return dfa;
}

/* ---- printing ----------------------------------------------------------- */
static std::string set_to_string(const std::set<int> &s) {
    std::string result = "{";
    bool first = true;
    for (int v : s) {
        if (!first) result += ", ";
        result += "q" + std::to_string(v);
        first = false;
    }
    result += "}";
    return result;
}

static void print_nfa(const NFA &nfa) {
    std::vector<std::string> symbols = {EPS};
    for (auto &s : nfa.alphabet) symbols.push_back(s);

    int col_w = 18;
    std::cout << std::left << std::setw(10) << "STATE";
    for (auto &sym : symbols)
        std::cout << "| " << std::setw(col_w) << sym;
    std::cout << "\n";

    int total_w = 10 + static_cast<int>(symbols.size()) * (col_w + 2);
    std::cout << std::string(total_w, '-') << "\n";

    for (int st = 0; st < nfa.num_states; st++) {
        std::string prefix;
        if (st == nfa.start) prefix += "->";
        if (nfa.accept_states.count(st)) prefix += "*";
        std::string label = prefix + "q" + std::to_string(st);
        std::cout << std::left << std::setw(10) << label;

        for (auto &sym : symbols) {
            auto it = nfa.trans.find({st, sym});
            std::string cell = "-";
            if (it != nfa.trans.end() && !it->second.empty()) {
                cell = "";
                bool first = true;
                for (int t : it->second) {
                    if (!first) cell += ", ";
                    cell += "q" + std::to_string(t);
                    first = false;
                }
            }
            std::cout << "| " << std::setw(col_w) << cell;
        }
        std::cout << "\n";
    }
    std::cout << std::string(total_w, '-') << "\n";
}

static void print_dfa(const DFA &dfa) {
    std::vector<std::string> symbols(dfa.alphabet.begin(), dfa.alphabet.end());

    int col_w = 10;
    int set_w = 26;
    std::cout << std::left << std::setw(10) << "STATE"
              << std::setw(set_w) << "NFA STATES";
    for (auto &sym : symbols)
        std::cout << "| " << std::setw(col_w) << sym;
    std::cout << "\n";

    int total_w = 10 + set_w + static_cast<int>(symbols.size()) * (col_w + 2);
    std::cout << std::string(total_w, '-') << "\n";

    for (auto &ds : dfa.states) {
        std::string prefix;
        if (ds.id == dfa.start) prefix += "->";
        if (ds.is_accept) prefix += "*";
        std::string label = prefix + "D" + std::to_string(ds.id);
        std::cout << std::left << std::setw(10) << label
                  << std::setw(set_w) << set_to_string(ds.nfa_states);

        for (auto &sym : symbols) {
            auto it = ds.transitions.find(sym);
            std::string cell = "-";
            if (it != ds.transitions.end())
                cell = "D" + std::to_string(it->second);
            std::cout << "| " << std::setw(col_w) << cell;
        }
        std::cout << "\n";
    }
    std::cout << std::string(total_w, '-') << "\n";
}

/* ---- build example NFA for (a|b)*abb ------------------------------------ */
static NFA build_example_nfa() {
    NFA nfa;
    nfa.start = 0;
    nfa.accept_states = {10};
    nfa.num_states = 11;

    struct { int from; std::string sym; int to; } transitions[] = {
        {0, EPS, 1}, {0, EPS, 7},
        {1, EPS, 2}, {1, EPS, 4},
        {2, "a", 3},
        {3, EPS, 6},
        {4, "b", 5},
        {5, EPS, 6},
        {6, EPS, 1}, {6, EPS, 7},
        {7, "a", 8},
        {8, "b", 9},
        {9, "b", 10},
    };

    for (auto &t : transitions)
        nfa.add_transition(t.from, t.sym, t.to);

    return nfa;
}

/* ---- main --------------------------------------------------------------- */
int main() {
    std::cout << "======================================================================\n"
              << "  NFA -> DFA (Subset Construction)  --  Exercise 3\n"
              << "======================================================================\n\n";

    std::cout << "Choose input mode:\n"
              << "  1) Use default NFA for (a|b)*abb\n"
              << "  2) Enter your own NFA\n"
              << "\nEnter choice [1]: ";

    std::string choice;
    std::getline(std::cin, choice);

    NFA nfa;
    if (choice == "2") {
        std::cout << "\nEnter number of NFA states: ";
        int n; std::cin >> n;
        nfa.num_states = n;

        std::cout << "Enter start state: ";
        std::cin >> nfa.start;

        std::cout << "Enter number of accept states: ";
        int na; std::cin >> na;
        std::cout << "Enter accept state IDs: ";
        for (int i = 0; i < na; i++) {
            int acc; std::cin >> acc;
            nfa.accept_states.insert(acc);
        }

        std::cout << "Enter number of transitions: ";
        int nt; std::cin >> nt;
        std::cout << "Enter transitions (from symbol to), use 'eps' for epsilon:\n";
        for (int i = 0; i < nt; i++) {
            int from, to;
            std::string sym;
            std::cin >> from >> sym >> to;
            nfa.add_transition(from, sym, to);
        }
    } else {
        nfa = build_example_nfa();
    }

    std::cout << "\n--- NFA Transition Table ---\n\n";
    print_nfa(nfa);

    DFA dfa = subset_construction(nfa);

    std::cout << "\n--- DFA Transition Table (Subset Construction) ---\n\n";
    print_dfa(dfa);

    /* summary */
    std::cout << "\nSummary:\n";
    std::cout << "  NFA states: " << nfa.num_states << "\n";
    std::cout << "  DFA states: " << dfa.states.size() << "\n";
    std::cout << "  DFA start : D" << dfa.start << "\n";
    std::cout << "  DFA accept: ";
    bool first = true;
    for (auto &ds : dfa.states) {
        if (ds.is_accept) {
            if (!first) std::cout << ", ";
            std::cout << "D" << ds.id;
            first = false;
        }
    }
    std::cout << "\n";

    return 0;
}
